### PR TITLE
cpu/lpc1768: add gpio peripheral (+ board adaptions)

### PR DESCRIPTION
**Pull request by basilfx**
I needed GPIO on the LPC1768 (more specific, on the Seeeduino Arch Pro), but I realized it wasn't there. Since I need it for a demo tomorrow, I decided to add it :-)

So far output mode works, but that's all I tested. Have to dig into interrupts, which seem to be a bit more complex.

Adapted `mbed_lpc1768` and `seeeduino-arch_pro` boards, including SAUL support for GPIO. I realize that they overlap a lot, but the boards are not quite similar.

WIP, needs a lot of cleanup.

### CODE REVIEW

**Comment by aabadie** [comment]
is this still works in progress @basilfx ?

File: boards/mbed_lpc1768/include/board.h [hunk 1]
```diff
@@ -41,24 +41,18 @@ extern "C" {
...
-#define LED1_ON             (LED_PORT->FIOSET  = LED1_MASK)
-#define LED1_OFF            (LED_PORT->FIOCLR  = LED1_MASK)
-#define LED1_TOGGLE         (LED_PORT->FIOPIN ^= LED1_MASK)
-#define LED2_ON             (LED_PORT->FIOSET  = LED2_MASK)
-#define LED2_OFF            (LED_PORT->FIOCLR  = LED2_MASK)
-#define LED2_TOGGLE         (LED_PORT->FIOPIN ^= LED2_MASK)
-#define LED3_ON             (LED_PORT->FIOSET  = LED3_MASK)
-#define LED3_OFF            (LED_PORT->FIOCLR  = LED3_MASK)
-#define LED3_TOGGLE         (LED_PORT->FIOPIN ^= LED3_MASK)
+#define LED0_ON             gpio_clear(LED0_PIN)
```
Comment (aabadie) [inline boards/mbed_lpc1768/include/board.h]: Looks like this is inverted.

**Comment by aabadie** [comment]
ping @basilfx

**Comment by basilfx** [comment]
@aabadie Sorry, I'm very busy with work-work. Will have to wait.

**Comment by aabadie** [comment]
ping @basilfx. Any chance to apply the requested changes for the next release ?

**Comment by aabadie** [comment]
@basilfx any chance for this one to be ready for the release ?

**Comment by basilfx** [comment]
@aabadie I finally found the time to fix this one.

Re-implemented the GPIO driver, added support for interrupts and fixed board LEDs.

File: boards/seeeduino_arch-pro/include/board.h [hunk 2]
```diff
@@ -42,24 +42,18 @@ extern "C" {
...
-#define LED1_ON             (LED_PORT->FIOCLR  = LED1_MASK)
-#define LED1_OFF            (LED_PORT->FIOSET  = LED1_MASK)
-#define LED1_TOGGLE         (LED_PORT->FIOPIN ^= LED1_MASK)
-#define LED2_ON             (LED_PORT->FIOCLR  = LED2_MASK)
-#define LED2_OFF            (LED_PORT->FIOSET  = LED2_MASK)
-#define LED2_TOGGLE         (LED_PORT->FIOPIN ^= LED2_MASK)
-#define LED3_ON             (LED_PORT->FIOCLR  = LED3_MASK)
-#define LED3_OFF            (LED_PORT->FIOSET  = LED3_MASK)
-#define LED3_TOGGLE         (LED_PORT->FIOPIN ^= LED3_MASK)
+#define LED0_ON             gpio_clear(LED0_PIN)
```
[the remark this answers is not in the document]
Comment (basilfx) [reply boards/seeeduino_arch-pro/include/board.h]: I prefer to keep implementation in one place (`gpio.c`), and not having two different ways to toggle a LED.

File: cpu/lpc1768/periph/gpio.c [hunk 3]
```diff
@@ -0,0 +1,238 @@
...
+    LPC_GPIO_TypeDef *base = _base(pin);
+
+    base->FIOCLR |= (1 << _pin(pin));
+}
+
+void gpio_toggle(gpio_t pin)
+{
+    LPC_GPIO_TypeDef *base = _base(pin);
+
+    if (base->FIOSET & _pin(pin)) {
```
[the remark this answers is not in the document]
Comment (basilfx) [reply cpu/lpc1768/periph/gpio.c]: Fixed.

File: boards/seeeduino_arch-pro/board.c [hunk 4]
```diff
@@ -35,24 +37,17 @@ void board_init(void)
...
- * soldered onto the board they are fixed to their CPU pins.
- *
- * The LEDs are connected to the following pins:
- * - LED1: P1.18
- * - LED2: P1.20
- * - LED3: P1.21
- * - LED4: P1.23
- *
- * The LEDs are active-low (current-sink).
+ * @brief Initialize the boards on-board LEDs.
```
[the remark this answers is not in the document]
Comment (basilfx) [reply boards/seeeduino_arch-pro/board.c]: Fixed.

File: boards/mbed_lpc1768/include/board.h (hunk 1 above)
Comment (basilfx) [reply boards/mbed_lpc1768/include/board.h]: Fixed.

On the Seeeduino Arch-Pro, they are inverted :-)

File: boards/mbed_lpc1768/board.c [hunk 5]
```diff
@@ -34,22 +36,17 @@ void board_init(void)
...
- *
- * The LED initialization is hard-coded in this function. As the LEDs are soldered
- * onto the board they are fixed to their CPU pins.
- *
- * The LEDs are connected to the following pins:
- * - LED1: P1.18
- * - LED2: P1.20
- * - LED3: P1.21
- * - LED4: P1.23
+ * @brief Initialize the boards on-board LEDs.
```
[the remark this answers is not in the document]
Comment (basilfx) [reply boards/mbed_lpc1768/board.c]: Fixed.

File: boards/seeeduino_arch-pro/Makefile.include [hunk 6]
```diff
@@ -11,6 +11,7 @@ include $(RIOTMAKE)/tools/serial.inc.mk
 DEBUG_ADAPTER ?= dap
 
 # this board uses openocd
+export IMAGE_FILE = $(HEXFILE)
```
Comment (aabadie) [inline boards/seeeduino_arch-pro/Makefile.include]: Looks unrelated. Maybe #8845 will help ?

Comment (basilfx) [reply boards/seeeduino_arch-pro/Makefile.include]: It's needed to flash the HEX file instead of the ELF file. #8845 would help, yes.

Comment (aabadie) [reply boards/seeeduino_arch-pro/Makefile.include]: I would prefer to make this change in a separate PR. Is it ok for you ?

Comment (basilfx) [reply boards/seeeduino_arch-pro/Makefile.include]: I'll put this in another PR.

Comment (basilfx) [reply boards/seeeduino_arch-pro/Makefile.include]: Created #8866.

**Comment by basilfx** [comment]
Rebased and squashed.

**Comment by basilfx** [comment]
Murdock is happy!
